Specialize `OsString::push` and `OsString as From` for UTF-8

thaliaarchi opened the pull request:
When concatenating two WTF-8 strings, surrogate pairs at the boundaries need to be joined. However, since UTF-8 strings cannot contain surrogate halves, this check can be skipped when one string is UTF-8. Specialize `OsString::push` to use a more efficient concatenation in this case.

The WTF-8 version of `OsString` tracks whether it is known to be valid UTF-8 with its `is_known_utf8` field. Specialize `From<AsRef<OsStr>>` so this can be set for UTF-8 string types.

Unfortunately, a specialization for `T: AsRef<str>` conflicts with `T: AsRef<OsStr>`, so stamp out string types with a macro.

r? @ChrisDenton

thaliaarchi committed:
Specialize OsString::push for strings

When concatenating two WTF-8 strings, surrogate pairs at the boundaries
need to be joined. However, since UTF-8 strings cannot contain surrogate
halves, this check can be skipped when one string is UTF-8. Specialize
`OsString::push` to use a more efficient concatenation in this case.

Unfortunately, a specialization for `T: AsRef<str>` conflicts with
`T: AsRef<OsStr>`, so stamp out string types with a macro.

bjorn3 commented:
> Avoid trying to join surrogates at boundaries when concatenating a WTF-8 OsString and a UTF-8 string.

Is this a performance optimization of a correctness fix?

thaliaarchi commented:
Performance. When the LHS ends with a leading surrogate and the RHS starts with a trailing surrogate, then they need to be joined. A WTF-8 `OsString` can contain surrogate halves, but a `&str` cannot since UTF-8 forbids them. Thus, concatenating the two will never join surrogates.

thaliaarchi committed:
Specialize constructing OsString from strings

The WTF-8 version of `OsString` tracks whether it is known to be valid
UTF-8 with its `is_known_utf8` field. Specialize `From<AsRef<OsStr>>` so
this can be set for UTF-8 string types.

thaliaarchi commented:
I have now pushed another similar specialization. Since the WTF-8 `OsString` tracks whether it is known to be UTF-8 with its `is_known_utf8` field, make sure `From<AsRef<OsStr>>` sets that to true when converting from UTF-8 strings.

joboet commented:
Great idea!
@bors r+

bors commented:
:pushpin: Commit 83407b828bcf5c883b2ab50b958edbcba0738f6a has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 83407b828bcf5c883b2ab50b958edbcba0738f6a -->
<!-- homu: {"type":"Approved","sha":"83407b828bcf5c883b2ab50b958edbcba0738f6a","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
